rtt/base/DataObjectLockFree.hpp

```diff
@@ -198,34 +237,77 @@ namespace RTT
              * not increment the counter on write_ptr+1). Hence, no
              * locking is needed.
              */
+
+            if (!initialized) {
+                types::TypeInfo *ti = types::Types()->getTypeInfo<DataType>();
+                log(Error) << "You set a lock-free data object of type " << (ti ? ti->getTypeName() : "(unknown)") << " without initializing it with a data sample. "
+                           << "This might not be real-time safe." << endlog();
+                data_sample(DataType(), true);
+            }
+
             // writeout in any case
-            write_ptr->data = push;
-            PtrType wrote_ptr = write_ptr;
+            PtrType writing = write_ptr;
+            writing->data = push;
+            writing->status = NewData;
+
             // if next field is occupied (by read_ptr or counter),
             // go to next and check again...
             while ( oro_atomic_read( &write_ptr->next->counter ) != 0 || write_ptr->next == read_ptr )
                 {
                     write_ptr = write_ptr->next;
-                    if (write_ptr == wrote_ptr)
-                        return; // nothing found, to many readers !
+                    if (write_ptr == writing)
+                        return false; // nothing found, to many readers !
                 }
 
             // we will be able to move, so replace read_ptr
-            read_ptr  = wrote_ptr;
+            read_ptr  = writing;
             write_ptr = write_ptr->next; // we checked this in the while loop
+            return true;
         }
 
-        virtual void data_sample( const DataType& sample ) {
-            // prepare the buffer.
-            for (unsigned int i = 0; i < BUF_LEN-1; ++i) {
-                data[i].data = sample;
-                data[i].next = &data[i+1];
+        virtual bool data_sample( const DataType& sample, bool reset = true ) {
+            if (!initialized || reset) {
+                // prepare the buffer.
+                for (unsigned int i = 0; i < BUF_LEN; ++i) {
+                    data[i].data = sample;
+                    data[i].status = NoData;
+                    data[i].next = &data[i+1];
+                }
+                data[BUF_LEN-1].next = &data[0];
+                initialized = true;
+                return true;
+            } else {
+                return initialized;
             }
-            data[BUF_LEN-1].data = sample;
-            data[BUF_LEN-1].next = &data[0];
+        }
+
+        // This is actually a copy of Get(), but it only sets the status to NoData once a valid buffer has been found.
+        // Subsequent read() calls will read from the same buffer and will return NoData until a new sample has been written.
+        virtual void clear() {
+            if (!initialized) return;
+
+            PtrType reading;
```

Comment:
Wouldn't it make sense to factor out this (complex) piece of logic into a acquireReadPointer or something like that ?